File: TerraNova.Client/NetworkClient.cs
using LiteNetLib;
using LiteNetLib.Utils;
using Microsoft.Extensions.Logging;
using TerraNova.Shared;

namespace TerraNova;

/// <summary>
/// Network client that connects to the game server and syncs world data
/// </summary>
public class NetworkClient : INetworkClient, INetEventListener
{
    private readonly NetManager _netManager;
    private readonly ILogger<NetworkClient> _logger;
    private NetPeer? _serverPeer;
    private World? _world;
    private bool _worldReceived = false;

    public bool IsConnected => _serverPeer != null && _serverPeer.ConnectionState == ConnectionState.Connected;
    public bool WorldReceived => _worldReceived;
    public World? World => _world;
    public bool WorldChanged { get; set; }

    public event Action<Vector2i, BlockData[]>? OnChunkReceived;
    public event Action<int, int, int, BlockType>? OnBlockUpdate;

    public NetworkClient(ILogger<NetworkClient> logger)
    {
        _logger = logger;
        _netManager = new NetManager(this);
    }

    public void Connect(string host, int port, string playerName)
    {
        _netManager.Start();
        _serverPeer = _netManager.Connect(host, port, "TerraNova");

        _logger.LogInformation("Connecting to {Host}:{Port}...", host, port);
    }

    public void Update()
    {
        _netManager.PollEvents();
    }

    public void Disconnect()
    {
        _netManager.Stop();
        _logger.LogInformation("Disconnected from server");
    }

    public void SendBlockUpdate(int x, int y, int z, BlockType blockType)
    {
        if (_serverPeer == null || !IsConnected)
        {
            _logger.LogWarning("Cannot send block update: not connected to server");
            return;
        }

        var blockUpdate = new BlockUpdateMessage(x, y, z, blockType);
        var writer = new NetDataWriter();
        writer.Put((byte)MessageType.BlockUpdate);
        writer.Put(blockUpdate);

        _serverPeer.Send(writer, DeliveryMethod.ReliableOrdered);
        _logger.LogDebug("Sent block update: ({X},{Y},{Z}) -> {BlockType}", x, y, z, blockType);
    }

    public void RequestChunks(Vector2i[] chunkPositions)
    {
        if (_serverPeer == null || !IsConnected)
        {
            _logger.LogWarning("Cannot request chunks: not connected to server");
            return;
        }

        var chunkRequest = new ChunkRequestMessage(chunkPositions);
        var writer = new NetDataWriter();
        writer.Put((byte)MessageType.ChunkRequest);
        writer.Put(chunkRequest);

        _serverPeer.Send(writer, DeliveryMethod.ReliableOrdered);
        _logger.LogDebug("Requested {Count} chunks from server", chunkPositions.Length);
    }

    // INetEventListener implementation
    public void OnPeerConnected(NetPeer peer)
    {
        _logger.LogInformation("Connected to server!");

        // Initialize with an empty world immediately
        _world = new World();
        _worldReceived = true;
        _logger.LogInformation("Empty world initialized - chunks will load on demand");

        // Send connection message
        var writer = new NetDataWriter();
        writer.Put((byte)MessageType.ClientConnect);
        writer.Put(new ClientConnectMessage("Player"));

        peer.Send(writer, DeliveryMethod.ReliableOrdered);
    }

    public void OnPeerDisconnected(NetPeer peer, DisconnectInfo disconnectInfo)
    {
        _logger.LogInformation("Disconnected from server: {Reason}", disconnectInfo.Reason);
        _serverPeer = null;
    }

    public void OnNetworkReceive(NetPeer peer, NetPacketReader reader, byte channelNumber, DeliveryMethod deliveryMethod)
    {
        var messageType = (MessageType)reader.GetByte();

        switch (messageType)
        {
            case MessageType.WorldData:
                var worldData = reader.GetWorldDataMessage();
                ReceiveWorldData(worldData);
                break;

            case MessageType.ChunkData:
                var chunkData = reader.GetChunkDataMessage();
                ReceiveChunkData(chunkData);
                break;

            case MessageType.BlockUpdate:
                var blockUpdate = reader.GetBlockUpdateMessage();
                _world?.SetBlock(blockUpdate.X, blockUpdate.Y, blockUpdate.Z, blockUpdate.NewType);
                OnBlockUpdate?.Invoke(blockUpdate.X, blockUpdate.Y, blockUpdate.Z, blockUpdate.NewType);
                _logger.LogInformation("Block updated at ({X},{Y},{Z})", blockUpdate.X, blockUpdate.Y, blockUpdate.Z);
                break;
        }

        reader.Recycle();
    }

    public void OnNetworkError(System.Net.IPEndPoint endPoint, System.Net.Sockets.SocketError socketError)
    {
        _logger.LogError("Network error: {SocketError}", socketError);
    }

    public void OnNetworkReceiveUnconnected(System.Net.IPEndPoint remoteEndPoint, NetPacketReader reader, UnconnectedMessageType messageType)
    {
        // Not used
    }

    public void OnNetworkLatencyUpdate(NetPeer peer, int latency)
    {
        // Not used
    }

    public void OnConnectionRequest(ConnectionRequest request)
    {
        // Client doesn't accept connections
        request.Reject();
    }

    private void ReceiveWorldData(WorldDataMessage worldData)
    {
        _logger.LogInformation("Received world data: {BlockCount} blocks", worldData.Blocks.Length);

        // Create world and populate it (legacy support for full world data)
        if (_world == null)
        {
            _world = new World();
        }

        foreach (var block in worldData.Blocks)
        {
            _world.SetBlock(block.X, block.Y, block.Z, block.Type);
        }

        _worldReceived = true;
        WorldChanged = true;
        _logger.LogInformation("World loaded from server!");
    }

    private void ReceiveChunkData(ChunkDataMessage chunkData)
    {
        _logger.LogInformation("Received chunk ({X},{Z}) with {BlockCount} blocks",
            chunkData.ChunkPosition.X, chunkData.ChunkPosition.Z, chunkData.Blocks.Length);

        if (_world == null)
        {
            _logger.LogWarning("Cannot receive chunk data: world not initialized");
            return;
        }

        // Add all blocks from this chunk to the world
        foreach (var block in chunkData.Blocks)
        {
            _world.SetBlock(block.X, block.Y, block.Z, block.Type);
        }

        // Notify listeners that a chunk was received
        OnChunkReceived?.Invoke(chunkData.ChunkPosition, chunkData.Blocks);
    }
}

File: TerraNova.Web/wwwroot/js/networkClient.js
// Terra Nova - WebSocket Network Client
// Handles connection to the game server via WebSocket

window.terraNovaNetwork = (function() {
    let ws = null;
    let connected = false;
    let worldDataCallback = null;
    let blockUpdateCallback = null;
    let chunkDataCallback = null;

    /**
     * Connect to the game server
     * @param {string} serverUrl - WebSocket server URL (e.g., "ws://localhost:5000/ws")
     * @param {function} onWorldData - Callback when world data is received
     * @param {function} onBlockUpdate - Callback when block update is received
     * @param {function} onChunkData - Callback when chunk data is received
     */
    function connect(serverUrl, onWorldData, onBlockUpdate, onChunkData) {
        console.log(`Connecting to server: ${serverUrl}`);

        worldDataCallback = onWorldData;
        blockUpdateCallback = onBlockUpdate;
        chunkDataCallback = onChunkData;

        ws = new WebSocket(serverUrl);

        ws.onopen = () => {
            console.log('WebSocket connected');
            connected = true;

            // Send client connect message
            const connectMsg = {
                type: MessageType.CLIENT_CONNECT,
                playerName: 'WebPlayer'
            };
            sendMessage(connectMsg);
        };

        ws.onmessage = (event) => {
            try {
                const message = JSON.parse(event.data);
                handleMessage(message);
            } catch (error) {
                console.error('Error parsing message:', error);
            }
        };

        ws.onerror = (error) => {
            console.error('WebSocket error:', error);
        };

        ws.onclose = () => {
            console.log('WebSocket disconnected');
            connected = false;
        };
    }

    /**
     * Disconnect from the server
     */
    function disconnect() {
        if (ws) {
            ws.close();
            ws = null;
            connected = false;
        }
    }

    /**
     * Send a message to the server
     * @param {object} message - Message object to send
     */
    function sendMessage(message) {
        if (ws && ws.readyState === WebSocket.OPEN) {
            ws.send(JSON.stringify(message));
        } else {
            console.warn('Cannot send message: WebSocket not connected');
        }
    }

    /**
     * Handle incoming message from server
     * @param {object} message - Parsed message object
     */
    function handleMessage(message) {
        switch (message.type) {
            case MessageType.WORLD_DATA:
                console.log(`Received world data: ${message.blocks.length} blocks`);
                if (worldDataCallback) {
                    worldDataCallback(message.blocks);
                }
                break;

            case MessageType.BLOCK_UPDATE:
                console.log(`Block update: (${message.x}, ${message.y}, ${message.z}) = ${message.blockType}`);

                // Update block interaction system's world data
                if (window.terraNovaBlockInteraction) {
                    window.terraNovaBlockInteraction.setBlockData(message.x, message.y, message.z, message.blockType);
                    window.terraNovaBlockInteraction.updateSelection(); // Refresh selection
                }

                if (blockUpdateCallback) {
                    blockUpdateCallback(message.x, message.y, message.z, message.blockType);
                }
                break;

            case MessageType.CHUNK_DATA:
                console.log(`Received chunk data: (${message.chunkX}, ${message.chunkZ}) with ${message.blocks.length} blocks`);
                if (chunkDataCallback) {
                    chunkDataCallback(message.chunkX, message.chunkZ, message.blocks);
                }
                break;

            default:
                console.warn('Unknown message type:', message.type);
        }
    }

    /**
     * Send a block update to the server
     * @param {number} x - Block X position
     * @param {number} y - Block Y position
     * @param {number} z - Block Z position
     * @param {number} blockType - Block type (0 = Air, 1 = Grass, etc.)
     */
    function sendBlockUpdate(x, y, z, blockType) {
        const message = {
            type: MessageType.BLOCK_UPDATE,
            x: x,
            y: y,
            z: z,
            blockType: blockType
        };
        sendMessage(message);
    }

    /**
     * Request specific chunks from the server
     * @param {Array<{x: number, z: number}>} chunkPositions - Array of chunk positions to request
     */
    function requestChunks(chunkPositions) {
        const message = {
            type: MessageType.CHUNK_REQUEST,
            chunkPositions: chunkPositions
        };
        sendMessage(message);
        console.log(`Requested ${chunkPositions.length} chunks from server`);
    }

    /**
     * Check if connected to server
     */
    function isConnected() {
        return connected;
    }

    // Expose public API
    return {
        connect: connect,
        disconnect: disconnect,
        sendBlockUpdate: sendBlockUpdate,
        requestChunks: requestChunks,
        isConnected: isConnected
    };
})();
